bpf/process/bpf_execve_event.c

```diff
@@ -329,19 +330,29 @@ execve_send(void *ctx)
 		memset(&curr->bin, 0, sizeof(curr->bin));
 #ifdef __LARGE_BPF_PROG
 		// read from proc exe stored at execve time
-		if (event->exe.len <= BINARY_PATH_MAX_LEN) {
+		if (event->exe.len <= BINARY_PATH_MAX_LEN && !event->exe.error) {
 			curr->bin.path_length = probe_read(curr->bin.path, event->exe.len, event->exe.off);
 			if (curr->bin.path_length == 0)
 				curr->bin.path_length = event->exe.len;
+		} else {
+			heap = map_lookup_elem(&execve_heap, &zero);
```

Comment:
so I'm not sure it's really appropriate as a fix:
- The current implementation is indeed limited to 256 bytes on `exe` read, but this could be increased (maybe that's what you want to actually do). This is indeed bad that it's failing open for something as Prefix.
- Here you read `execve_heap` that contains the binary only because of the way we read args at `tracepoint/sys_execve`, I don't think there are any guarantees that this heap will actually contain this btw. We skip the binary name that is the first args, but this rollbacks to the old situation where we retrieved the filename from that args which could lead to read a relative path or a symlink (we switched to exe for that reason, reading the syscall args wasn't reliable). See: https://github.com/cilium/tetragon/blob/92a17c2386780497a5541dae5d0b0824f036dc0c/bpf/process/bpf_execve_event.c#L248-L254

I agree that there is a limitation on the current implementation which is pretty explicit (with `BINARY_PATH_MAX_LEN`), but falling back to the way we do things on <5.4 kernels will not help as it's also flawed and can be escaped with relative paths symlinks.

And since you `probe_read_str` using `BINARY_PATH_MAX_LEN` that is set to 256, I'm not sure to understand how that helps?

We have unit tests on the function that uses `read_exe` and the error should be filled if the path is too big to enter the buffer of 256 (it seems that the limit is 255) so we should be aware that we could not match properly.

Reply:
Thinking a bit more about this, it's indeed not great that we never check for the error at read from https://github.com/cilium/tetragon/pull/1926 and that fallback to the `args`' filename could be an idea but it looks flawed.  Indeed, we have unit tests on the function that uses `read_exe` and the error should be filled if the path is too big to enter the buffer of 256 (it seems that the limit is 255) so we should be aware that we could not match properly.

Because as your test highlights, an errored dentry will not work since it's reading from the end to the start, and thus the beginning of the path containing the prefix will not be here. Switching back to the old implem (as it seems you do even though you read from the execve_heap instead of p->args) could work until you don't use any relative path or symlink.

So I see two things:
- support the error more explicitly and return an error to the user so that he knows we might have skipped an event because of path len limitation
- fix the issue by increasing `BINARY_PATH_MAX_LEN` to 4096.

Reply:
Thanks for detailed review! Yes you are right, relative paths and symlinks might be a problem. It is good to find a better solution that I proposed. Let me explain what I've done.

IIUC, we read the first argument here.
https://github.com/cilium/tetragon/blob/1a7b15abb3d48802f7f44d44a1c2c7cb58ec51f7/bpf/process/bpf_execve_event.c#L59-L69

Here we think that p->args contains the first argument, but it seems to me that p->args contains arguments starting from the second one. So that's why I changed the logic for old kernels.
https://github.com/cilium/tetragon/blob/1a7b15abb3d48802f7f44d44a1c2c7cb58ec51f7/bpf/process/bpf_execve_event.c#L338

Reply:
> So I see two things:
support the error more explicitly and return an error to the user so that he knows we might have skipped an event because of path len limitation
fix the issue by increasing BINARY_PATH_MAX_LEN to 4096.

I think it's a good idea to increase BINARY_PATH_MAX_LEN to 4096.  If so we don't  need `file_copy_reverse` here. And also, we don't need to have extra variables to hold postfix path.

Do you see any limitations that can stop us to increase the BINARY_PATH_MAX_LEN?
IIUC, limitations with prefix/postfix matching we can avoid by copying exact amount of data (STRING_PREFIX_MAX_LENGTH / STRING_POSTFIX_MAX_LENGTH).

Reply:
We can have problems here, if  __LARGE_MAP_KEYS is not defined, I think:

https://github.com/cilium/tetragon/blob/1a7b15abb3d48802f7f44d44a1c2c7cb58ec51f7/bpf/process/types/basic.h#L1581-L1587.

But we can have a check, for this case.

Reply:
> Here we think that p->args contains the first argument, but it seems to me that p->args contains arguments starting from the second one. So that's why I changed the logic for old kernels.

It's not the case, `event->process.args` is read in function `read_path` and the first string separated by `\0` is the first args, thus the binary given to sys_execve:

https://github.com/cilium/tetragon/blob/8e07d6cee05cda5f77a0191ca9250da5dc7c0b40/bpf/process/bpf_execve_event.c#L102-L111

Reply:
read_args, reads the args (hear the args of the execve binary not the args of the syscall), and thus use probe_read_str to remove the first part of the args, but does not touch it.

Reply:
Hmm... I see, thanks making it clear!

Reply:
> Do you see any limitations that can stop us to increase the BINARY_PATH_MAX_LEN?

It all boils down to make `prepend_name` support more than 256 bytes. I'm looking into it at the moment.